### PR TITLE
Refactor system_libs.py for easier combinations

### PR DESCRIPTION
**Pull request by quantum5**

### CODE REVIEW

File: embuilder.py [hunk 1]
```diff
@@ -235,115 +178,17 @@ def is_flag(arg):
         print('Skipping building of native-optimizer; EMSCRIPTEN_NATIVE_OPTIMIZER set in .emscripten config.')
       else:
         tasks += ['native_optimizer']
+    tasks += ['struct_info']
```
Comment (kripken) [inline embuilder.py]: why is this needed?

File: tests/test_other.py [hunk 2]
```diff
@@ -4630,7 +4630,7 @@ def test_only_force_stdlibs_2(self):
   }
 }
 ''')
-    with env_modify({'EMCC_FORCE_STDLIBS': 'libc,libc++abi,libc++,libdlmalloc,libpthreads_stub', 'EMCC_ONLY_FORCED_STDLIBS': '1'}):
+    with env_modify({'EMCC_FORCE_STDLIBS': 'libc,libc++abi,libc++,libmalloc,libpthreads', 'EMCC_ONLY_FORCED_STDLIBS': '1'}):
```
Comment (kripken) [inline tests/test_other.py]: what is `libmalloc`? in embuilder.py help I see `libdlmalloc` as I'd expect. are the names not coordinated between the two places?

also, why is `pthreads_stub` changing to `pthreads` here? that's a different library, or did the meaning change?

File: tools/system_libs.py [hunk 3]
```diff
@@ -78,161 +78,299 @@ def create_lib(libname, inputs):
...
+    path_components=['system', 'lib', 'libc', 'musl', 'src', 'string'],
+    filenames=['memset.c', 'memmove.c'])
+  other_files = files_in_path(
+    path_components=['system', 'lib', 'libc'],
+    filenames=['emscripten_memcpy.c'])
+  return math_files + string_files + other_files
+
+
+class Library(object):
+  # A name that is not None means a concrete library and not an abstract one.
```
Comment (kripken) [inline tools/system_libs.py]: what is an "abstract library"?

File: tools/system_libs.py [hunk 4]
```diff
@@ -78,161 +78,299 @@ def create_lib(libname, inputs):
...
+  name = None
+  depends = []
+  symbols = set()
+  js_depends = []
+
```
Comment (kripken) [inline tools/system_libs.py]: would be good to add some comments to explain what these fields are

File: tools/system_libs.py [hunk 5]
```diff
@@ -78,161 +78,299 @@ def create_lib(libname, inputs):
...
+    return self.get_base_name() + '.' + self.get_ext()
+
+  def get_symbols(self):
+    return self.symbols.copy()
+
+  def get_depends(self):
+    return self.depends
+
+  @classmethod
+  def variations(cls):
```
Comment (kripken) [inline tools/system_libs.py]: I assume this means variations like a `-mt`, etc.? would be good to add a comment explaining that

File: tests/test_other.py (hunk 2 above)
Comment (quantum5) [reply tests/test_other.py]: In the new system, we use simple names to refer to libraries. Depending on the current setting, it will link in the correct version of the library.

For example, `libmalloc` links in `dlmalloc` or `emmalloc` depending on what's specified in `Settings`. Similarly, `libpthreads` links in the actual library or the stub version.

File: embuilder.py (hunk 1 above)
Comment (quantum5) [reply embuilder.py]: `generated_struct_info.json` is no longer automatically built due to the build system change, so we make this task run explicitly when building all libraries.

File: tools/system_libs.py (hunk 3 above)
Comment (quantum5) [reply tools/system_libs.py]: An abstract subclass of `Library` that's meant to describe some feature common to many libraries, such as `MTLibrary`.

File: tools/system_libs.py (hunk 4 above)
Comment (quantum5) [reply tools/system_libs.py]: Done.

File: tools/system_libs.py (hunk 5 above)
Comment (quantum5) [reply tools/system_libs.py]: Renamed this to be more clear.

File: embuilder.py [hunk 6]
```diff
@@ -191,6 +132,8 @@ def main():
   def is_flag(arg):
     return arg.startswith('--')
 
+  shared.check_sanity()
```
Comment (kripken) [inline embuilder.py]: i vaguely remember this was here for a reason - that we must check here so the cache is invalidated up front, rather than later? please add a comment explaining this. also, I think this should run even without main - how about putting it at the top level at the bottom of the file before checking if we are in a main program?

File: embuilder.py (hunk 1 above)
Comment (kripken) [reply embuilder.py]: please add a comment explaining that, as otherwise it's very surprising to see it singled out for special treatment. something like "is necessary for building anything else". but that's not the full explanation, since isn't libc in a similar situation?

File: tools/system_libs.py [hunk 7]
```diff
@@ -286,325 +599,392 @@ def create_libc(libname):
...
+    base = 'lib%s' % self.malloc
+
+    extra = ''
+    if self.is_debug:
+      extra += '_debug'
+    if not self.use_errno:
+      # emmalloc doesn't actually use errno, but it's easier to build it again
+      extra += '_noerrno'
+    if self.is_mt:
+      extra += '_threadsafe'
```
Comment (kripken) [inline tools/system_libs.py]: doesn't inheriting from `MTLibrary` handle adding the `-mt` suffix for us?

if this is overriding just to keep the same names as before, i think it's ok to change the name if it simplifies the code (which it would if just one place adds `-mt` or such to the name).

Comment (quantum5) [reply tools/system_libs.py]: Somewhat unfortunately, due to the way inheritance work, we cannot actually call `super` if we want use the prefix `libdlmalloc` or `libemmalloc`, which is the reason why we are handling it here manually.

File: embuilder.py (hunk 1 above)
Comment (quantum5) [reply embuilder.py]: It might make more sense for this to be listed in system targets.

File: tools/system_libs.py (hunk 7 above)
Comment (kripken) [reply tools/system_libs.py]: i probably don't understand yet. why doesn't this problem happen with all the other uses of MTLibrary?

Comment (quantum5) [reply tools/system_libs.py]: All other uses of `MTLibrary` are fine because they end up calling `Library`'s `get_base_name`, returning the `name` class attribute. However, the `name` attribute on `libmalloc` is `libmalloc`, so using inheritance, we'd be forced to have `libmalloc-mt-dlmalloc...`, which looks terrible.

Comment (kripken) [reply tools/system_libs.py]: I see, thanks.

**Comment by kripken** [comment]
Very nice work!

Overall I think this is good enough to land after the last changes, and we can iterate on improvements from there, so lgtm from me. However, since this is a big change and the others are away, let's wait til next week and see if we can get feedback from some of them.

File: ChangeLog.md [hunk 8]
```diff
@@ -19,6 +19,22 @@ See docs/process.md for how version tagging works.
...
+       and certain system libraries that are always built. For example, running
+       `embuilder.py build libc` no longer builds `libcompiler_rt` if it hasn't be built.
+     - When using `EMCC_FORCE_STDLIBS` with a list of libraries, you must now use
+       the simplified names, for example, `libmalloc` and `libpthreads` instead of
+       `libdlmalloc` or `libpthreads_stub`. These names will link in the correct
+       version of the library: if the build is configured to use `emmalloc`, `libmalloc`
+       will mean `libemmalloc`, and if thread support is disabled, `libpthreads` will
+       mean `libpthreads_stub`. This allows you to say `libmalloc` or `libpthreads` without
+       worrying about which implementation is supposed to be used, and avoid duplicate
+       symbols if you used the wrong implementation.
```
Comment (juj) [inline ChangeLog.md]: The above comment only pertains to `EMCC_FORCE_STDLIBS` and not to `embuilder.py` command line usage, right?

One core use case of embuilder.py is to drive a prebuild of specific system libraries at compiler toolchain package authoring time, so that these system libraries will not be attempted to be built on demand: this helps give fast first time experience to users, e.g. with Unity when they export their first HTML5 project, on a weak laptop building the system libraries would add several minutes to first build - leading users to perceive that web exports are really slow. Also another reason is that installation is done in 'Program Files' folder, where everything is read-only from the get-go. Building content on demand inside 'Program Files' would not work.

To achieve the above, we use embuilder.py to prebuild libemmalloc and libdlmalloc for example, so that still works? (Briefly looking it appears to me that it does, but just verifying)

**Comment by juj** [comment]
Hmm, there's some conflict upcoming with this PR I'm hoping can land soon https://github.com/emscripten-core/emscripten/pull/7672 . I hope that's easy to factor in to the structure that this PR brings?

File: tools/system_libs.py (hunk 7 above)
Comment (kripken) [reply tools/system_libs.py]: Thinking on this some, I'm not sure malloc is that special, and this issue may happen more over time. If inheritance is clumsy here, perhaps an option is to avoid it? That is, instead of
```
class SomeLibrary(MTLibrary, OtherLibrary):
  ..
```
there could be a single class for all library types, and "Modifiers" (VariantGenerators? Variators?) that are added to a list,
```
class SomeLibrary(Library):
  modifiers = [MTLibrary, OtherLibrary]
  ..
```
then the "Modifiers" would be called in the proper places?

Comment (quantum5) [reply tools/system_libs.py]: Actually now that I think about it, we can still do this with inheritance. For this use case, we just need a method that returns `self.name` that can be overridden. I suppose we might as well do this for `libmalloc` as an example for the future.

File: ChangeLog.md (hunk 8 above)
Comment (quantum5) [reply ChangeLog.md]: Yes, `embuilder.py` is meant to be able to build any version of the library you want. `EMCC_FORCE_STDLIBS` now uses the right variant for the current build.

File: tools/system_libs.py [hunk 9]
```diff
@@ -78,161 +78,480 @@ def create_lib(libname, inputs):
...
+
+    This is the dynamic version of `depends`.
+    """
+    return self.depends
+
+  @classmethod
+  def vary_on(cls):
+    """
+    Returns a list of boolean constructor arguments that defines the variations
+    of this library.
```
Comment (tlively) [inline tools/system_libs.py]: It would be good to clarify that these are strings. When I initially read this I thought "boolean contructor arguments" meant bools.

File: tools/system_libs.py [hunk 10]
```diff
@@ -78,161 +78,480 @@ def create_lib(libname, inputs):
...
+    vary_on = cls.vary_on()
+    return [dict(zip(vary_on, toggles)) for toggles in
+            itertools.product([False, True], repeat=len(vary_on))]
+
+  @classmethod
+  def get_default_variation(cls, **kwargs):
+    """
+    Construct the variation suitable for the current invocation of emscripten.
+
+    Subclasses should pass the keyword arguments introduced by itself to the
```
Comment (tlively) [inline tools/system_libs.py]:
```suggestion
    Subclasses should pass the keyword arguments they introduce to the
```

File: tools/system_libs.py [hunk 11]
```diff
@@ -286,325 +605,391 @@ def create_libc(libname):
...
+  cflags = ['-O2', '-fno-builtin']
+  src_dir = ['system', 'lib', 'libc', 'musl', 'src', 'math']
+  src_files = ['cos.c', 'cosf.c', 'cosl.c', 'sin.c', 'sinf.c', 'sinl.c',
+               'tan.c', 'tanf.c', 'tanl.c', 'acos.c', 'acosf.c', 'acosl.c',
+               'asin.c', 'asinf.c', 'asinl.c', 'atan.c', 'atanf.c', 'atanl.c',
+               'atan2.c', 'atan2f.c', 'atan2l.c', 'exp.c', 'expf.c', 'expl.c',
+               'log.c', 'logf.c', 'logl.c', 'pow.c', 'powf.c', 'powl.c']
+
+  def can_use(self):
+    # if building to wasm, we need more math code, since we have less builtins
```
Comment (tlively) [inline tools/system_libs.py]:
```suggestion
    # if building to wasm, we need more math code, since we have fewer builtins
```

File: tools/system_libs.py [hunk 12]
```diff
@@ -78,161 +78,480 @@ def create_lib(libname, inputs):
...
-      return ['-D__EMSCRIPTEN_OFFSCREEN_FRAMEBUFFER__']
-    else:
-      assert '-ofb' not in libname
-      return []
 
-  # libc
-  def create_libc(libname):
-    logger.debug(' building libc for cache')
+class CXXLibrary(Library):
+  emcc = shared.EMXX
```
Comment (tlively) [inline tools/system_libs.py]: This is very elegant :+1:

File: tools/system_libs.py [hunk 13]
```diff
@@ -286,325 +605,391 @@ def create_libc(libname):
...
-    o = in_temp(out_name)
-    cflags = ['-O2', '-fno-builtin']
-    if shared.Settings.USE_PTHREADS:
-      cflags += ['-s', 'USE_PTHREADS=1']
-    if shared.Settings.EMSCRIPTEN_TRACING:
-      cflags += ['--tracing']
-    if shared.Settings.DEBUG_LEVEL >= 3:
+    self.is_debug = kwargs.pop('is_debug')
+    self.use_errno = kwargs.pop('use_errno')
+    self.is_tracing = kwargs.pop('is_tracing')
```
Comment (tlively) [inline tools/system_libs.py]: What would happen if some abstract library wanted to use the same kwargs? Should these have a `malloc_` prefix or something like that?

File: tools/system_libs.py [hunk 14]
```diff
@@ -286,325 +605,391 @@ def create_libc(libname):
...
+  name = 'libubsan_rt_wasm'
+  depends = ['libsanitizer_common_rt_wasm']
+
+  includes = [['system', 'lib', 'compiler-rt', 'lib']]
+  cflags = ['-std=c++11', '-DUBSAN_CAN_USE_CXXABI']
+  src_dir = ['system', 'lib', 'compiler-rt', 'lib', 'ubsan']
+  src_glob = '*.cc'
+
+
+def calculate(temp_files, in_temp, stdout_, stderr_, forced=[]):
```
Comment (tlively) [inline tools/system_libs.py]: This may not be appropriate for this PR, but it would be great if this function had a more descriptive name.

File: tools/system_libs.py (hunk 7 above)
Comment (tlively) [reply tools/system_libs.py]: IMO using multiple inheritance makes the implementation harder to follow but makes the resulting client code more elegant. The code is documented well enough that I don't think the complexity of the implementation will be a maintenance issue going forward, so I'm happy with the current design.

@kripken I agree that your proposed solution would have a simpler implementation, so if you feel strongly about this then I would be happy going in that direction as well.

File: tools/system_libs.py [hunk 15]
```diff
@@ -78,161 +78,480 @@ def create_lib(libname, inputs):
...
+  # Static linking is tricky with LLVM, since e.g. memset might not be used
+  # from libc, but be used as an intrinsic, and codegen will generate a libc
+  # call from that intrinsic *after* static linking would have thought it is
+  # all in there. In asm.js this is not an issue as we do JS linking anyhow,
+  # and have asm.js-optimized versions of all the LLVM intrinsics. But for
+  # wasm, we need a better solution. For now, make another archive that gets
+  # included at the same time as compiler-rt.
+  # Note that this also includes things that may be depended on by those
+  # functions - fmin uses signbit, for example, so signbit must be here (so if
+  # fmin is added by codegen, it will have all it needs).
```
[the remark this answers is not in the document]
Comment (quantum5) [reply tools/system_libs.py]: This wasn't my code. I just moved this function outside of `calculate`.

File: tools/system_libs.py (hunk 11 above)
Comment (quantum5) [reply tools/system_libs.py]: This wasn't my comment either, but whatever.

File: tools/system_libs.py (hunk 13 above)
Comment (quantum5) [reply tools/system_libs.py]: That would only be a problem if `libmalloc` inherits from that abstract library. If that were the case, we probably wouldn't need this line.

If we really want to, we _could_ use the `__` name mangling, but I don't feel it's necessary.

**Comment by sbc100** [comment]
I'm a little concerned that we may have lost the ability to control the order of libraries. Were they in declaration order before and now alphabetical?   I am a little surprised that alphabetical order is sufficient but the tests pass so I guess its OK so far.

I suppose really we want to build a dependency DAG such that any dependent libraries always come after on the command line.   But I feel like the complexity here is already getting out of control.

**Comment by quantum5** [comment]
@sbc100 The actual order of libraries we pass to `wasm-ld` is handled by `add_library`, which adds the dependencies of the current library if it hasn't already been added to the command line. This behaviour did not change compared to before.

The thing that changed is that we now try to resolve the undefined symbols in our object files by going through our libraries in alphabetical order, instead of the declaration order from before.

Even before this change, the libraries were not linked in topological order based on dependencies. Compiling a C++ hello world program **before** this PR passes the libraries on the command line in this order:

```
~/.emscripten_cache/wasm-obj/libc++_noexcept.a
~/.emscripten_cache/wasm-obj/libc++abi.a
~/.emscripten_cache/wasm-obj/libc.a
~/.emscripten_cache/wasm-obj/libcompiler_rt.a
~/.emscripten_cache/wasm-obj/libc-wasm.a
~/.emscripten_cache/wasm-obj/libdlmalloc.a
~/.emscripten_cache/wasm-obj/libpthreads_stub.a
~/.emscripten_cache/wasm-obj/libcompiler_rt_wasm.a
~/.emscripten_cache/wasm-obj/libc_rt_wasm.a
```

We note clearly that `libpthreads_stub` and `libcompiler_rt` are declared to be dependent on `libc`, yet there are listed *after* `libc`, violating topological order.

If the libraries were supposed to be passed in topological order, we can add a topological sort. However, `libcompiler_rt` and `libc` actually depend on each other, both before and after this PR, so we aren't dealing with a DAG.

**Comment by quantum5** [comment]
Checking the man page for [`ld.lld(1)`](https://manpages.debian.org/unstable/lld-8/ld.lld-8.1.en.html), it appears that `lld` keeps track of previously seen symbols and links them in if a subsequent library depends on it, so it doesn't actually matter which order we specify the `.a` files on the command line.

**Comment by sbc100** [comment]
The order matters when one two libraries both define the same symbol.  The first one on the command line will generally win.  I'm not sure we take advantage of that anywhere.

The order in which `add_library` gets call does change under the new scheme doesn't it?  Because `system_libs` used to be an array that was manually constructed and now its built from the inheritance graph and then `sort`ed?    For example, previously `libhtml5` would always come before `libgl` simply becasue of the way they were listed.     Not I'm pretty sure `libgl` would come first because its alphabetical. 

As you say not a huge deal probably, especially for `lld`.   Note that the "linking" we implement for fastcomp doesn't ahve the ability to look back so order matters there more, but only for `.a` and not for `.bc` so again it doesn't matter much.

I guess the only time it will really matter is if we want to have a library that overrides symbols from another library.  Until that time maybe we can ignore this issue.  In that case it would be "depends" so much as "comes_before" maybe?

**Comment by quantum5** [comment]
If we do find ourselves needing the libraries to appear in certain orders, we could actually support having the libraries appear in the order they appear in `system_libs.py` by using a metaclass to keep track of the definition order, and then sorting based on that instead of the name.
